### PR TITLE
[NNX] NNX migration prep (7/N): NNX-native MaxEngine inference

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map

1. ✅ Add NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. Linen workflow unchanged. (PR #3427)
2. ✅ NNX sharding utilities: `get_abstract_state_nnx`, `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, `get_mesh_from_config`. (PR #3470)
3. ✅ NNX fully supported end-to-end: `TrainStateNNX`, model creation, gradient accumulation, checkpointing, and training loop dispatch. (PR #3500)
4. ✅ Sharding diagnostics on NNX, plus post-training bugfixes that surfaced once the NNX path got exercised end-to-end. (PR #3652)
4.5. ✅ Linen↔NNX checkpoint converter. (PR #3843)
4.6. ❌ Linen↔NNX checkpoint comparator (sibling branch on PR4.5).
5. ✅ NNX correctness fixes, feature enablements, and vocab tiling on NNX.
6. ✅ NNX-native DPO.
7. 🔄 **[This PR]** NNX-native MaxEngine inference. Drops the route-to-Linen path in `maxengine.py`; `pure_nnx=True` now drives a real NNX flow (two-mode abstract `Transformer`, `nnx.merge`/`nnx.state` per JIT body, pure-dict cache adapter so `bulk_insert` etc. reuse Linen plumbing). Stacks on PR6. Carve-outs (NotImplementedError'd with PR pointers): NNX+quantization → PR9.5; NNX+LoRA → PR8; `prefill_multisampling`/`prefill_concat`/`stack_prefill_result_cache=True` → follow-up.
8. ❌ NNX-native LoRA + GRPO.
9. ❌ NNX-aware QK-Clip + remaining checkpoint utilities.
9.5. ❌ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix.
10. ❌ Vocab tiling `custom_vjp` for NNX.
11. ❌ Set NNX defaults to `True`; regenerate sharding goldens; flip back integration-test `pure_nnx=False` annotations.
12. ❌ Delete Linen-specific code paths and NNX compatibility flags.

## Description

PR5 routed `maxengine.py` inference to the Linen implementation regardless of `pure_nnx`, so `pure_nnx=True` users silently got the Linen engine. This PR replaces that route with a real NNX flow.

When `config.pure_nnx=True`, the engine builds an NNX `Transformer`, splits it into `(params, cache, rest)` with `nnx.split`, and at every JIT body reconstructs the model with `nnx.merge(..., copy=True)` to run the forward pass. Linen is preserved byte-for-byte: every NNX edit is gated `if config.pure_nnx:`, and `pure_nnx=False` stays the default.

Diff: **+404 / −33 across 3 files**. The only deletion is the "use Linen path regardless of `pure_nnx`" comment block in `MaxEngine.__init__`.

## Design notes

- **Two abstract models.** `__init__` builds `self.model` (PREFILL, batch 1) and `self.model_ar` (AUTOREGRESSIVE, decode-state batch). NNX cache vars take their logical axis names (`CACHE_BATCH` vs `CACHE_BATCH_PREFILL`) from the construction `model_mode`; the decode-state cache must use AR-mode names so `bulk_insert`'s `cache_batch` substring lookup hits. `nnx.eval_shape` is called directly (not via `create_nnx_abstract_model`) to avoid the `jax.set_mesh` wrap that trips Flax 0.12.6 on logical-only axes like `"norm"`.
- **Cache adapter.** The engine's cache helpers (`bulk_insert`, `_insert_jit`, `_maybe_*_prefill_result_cache`) walk the cache by `path[-1].key`. NNX `nnx.Cache` state would expose a `.value` accessor there, so the cache flows as a plain dict on both paths: `nnx.State.to_pure_dict()` after the run, `nnx.replace_by_pure_dict()` before `nnx.merge`. The Linen helpers stay unchanged.
- **Param loading** (`_load_params_nnx`): accepts user-provided NNX params or loads via `from_pretrained`, materializes the `rest` (RNG) state once for `nnx.merge`, and builds the prefill/AR KV-cache shardings.
- **Annotation helpers** (`maxtext_utils.py`): `get_prefill_kv_cache_annotations_nnx` / `get_kv_cache_annotations_nnx` mirror the Linen helpers' return shape (a tree of `PartitionSpec`).

## Carve-outs

These raise `NotImplementedError` on the NNX path — deliberate scope cuts, not silent fallbacks. Each points at its follow-up; the Linen path (`pure_nnx=False`) remains the workaround.

| Site | Tracked in |
|------|------------|
| quantization (`quantize_params`, quantized-checkpoint load) | PR9.5 |
| LoRA (`load_single_adapter`) | PR8 |
| `prefill_multisampling`, `prefill_concat`, `stack_prefill_result_cache=True` | PR7.5 |

## Tests
```
$ pytest tests/integration/maxengine_test.py
$ pytest tests/unit/maxengine_test.py -v
```

### Main Branch
- Linen:
    - Terminal 1: https://paste.googleplex.com/5080929906327552
    - Terminal 2: https://paste.googleplex.com/6165057028358144
    - xprof: https://xprof.corp.google.com/memory_profile/wanglance-1375759241400258495

- NNX: Not working

### feat/nnx-native-maxengine Branch
- Linen:
    - Terminal 1: https://paste.googleplex.com/5055968814694400
    - Terminal 2: https://paste.googleplex.com/5981768359477248
    - xprof: https://xprof.corp.google.com/memory_profile/wanglance-4754692464398927922

- NNX:
    - Terminal 1: https://paste.googleplex.com/5633097478569984
    - Terminal 2: https://paste.googleplex.com/4783748351655936
    - xprof: https://xprof.corp.google.com/memory_profile/wanglance-7835295661845546410

## Inference Microbenchmark Test
- Main Branch vs feat/nnx-native-maxengine Branch: https://diff.googleplex.com/#key=LuIFJ9UhpQfW
- feat/nnx-native-maxengine Linen vs NNX: https://diff.googleplex.com/#key=vqIrjB38YgoX

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by bvandermoon — COMMENTED** [review]
Thank you @ecnal-cienet. It's generally looking good to me, but could you please run some manual maxengine/jetstream tests to confirm functionality too?

File: src/maxtext/inference/maxengine/maxengine.py [hunk 1]
```diff
@@ -282,11 +370,80 @@ def load_params(self, *args, params=None, rng: PRNGKeyType | None = None, **kwar
...
+          params_abs,
+          is_leaf=lambda x: isinstance(x, nnx.Variable),
+      )
+      params_state = jax.device_put(params, target_shardings)
+      # Build a concrete model once to capture a real `rest` (RNG vars) for nnx.merge.
+      # Wasteful but simple — the from_pretrained branch below avoids this.
+      with nn_partitioning.axis_rules(self.config.logical_axis_rules):
+        concrete_model = self._create_model_fn()
+      graphdef, _, _, rest_state = nnx.split(concrete_model, nnx.Param, nnx.Cache, ...)
+      self.graphdef = graphdef
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py:397]: Fixed

File: src/maxtext/inference/maxengine/maxengine.py [hunk 2]
```diff
@@ -282,11 +370,80 @@ def load_params(self, *args, params=None, rng: PRNGKeyType | None = None, **kwar
...
+    """NNX equivalent of load_params: returns an nnx.Param state and populates KV cache shardings."""
+    if self.model.quant is not None:
+      raise NotImplementedError("pure_nnx + quantization not yet supported. Use pure_nnx=False.")
+
+    if params:
+      print("Resharding given NNX params")
+      _, params_abs, _ = nnx.split(self.model, nnx.Param, ...)
+      target_shardings = jax.tree.map(
+          lambda x: x.sharding if hasattr(x, "sharding") else None,
+          params_abs,
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py]: Fixed

**Comment by ecnal-cienet** [comment]
HI @bvandermoon,
I added Maxengine manual test result on PR body.
Thank you.

**Comment by bvandermoon** [comment]
> HI @bvandermoon, I added Maxengine manual test result on PR body. Thank you.

Awesome, thank you @ecnal-cienet. I see the HLOs are identical as expected

File: src/maxtext/inference/maxengine/maxengine.py [hunk 3]
```diff
@@ -110,10 +114,33 @@ def __init__(self, config: Any, devices: Any | None = None):
...
+      # the jax.set_mesh wrap that trips Flax 0.12.6 on logical-only axes like "norm".
+      _create_model = model_creation_utils.get_nnx_create_model_fn(config, mesh=self._mesh, model_mode=MODEL_MODE_PREFILL)
+      _create_model_ar = model_creation_utils.get_nnx_create_model_fn(
+          config, mesh=self._mesh, model_mode=MODEL_MODE_AUTOREGRESSIVE
+      )
+      with nn_partitioning.axis_rules(config.logical_axis_rules):
+        abstract_model = nnx.eval_shape(_create_model)
+        abstract_model_ar = nnx.eval_shape(_create_model_ar)
+      self.model = abstract_model
+      self.model_ar = abstract_model_ar
```
Comment (bvandermoon) [inline src/maxtext/inference/maxengine/maxengine.py:131]: Does initializing both `model` and `model_ar` take up additional memory? I don't see that in the profile you shared, but it seems like it would to me

Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py:131]: Hi @bvandermoon,
I double checked the code and there isn't actually extra memory cost here. Both `self.model` and `self.model_ar` are built via `nnx.eval_shape(...)`, which only traces the model abstractly: each leaf comes back as a jax.ShapeDtypeStruct, not a real array. So no HBM allocated.

The actual param allocation happens once in `_load_params_nnx` (either `device_put` of user-provided params or `from_pretrained` of a checkpoint), and we use `create_nnx_sharded_model` so the params are born sharded — peak is ≈1× weights, not 2×.

The two abstract models exist purely so the NNX `Cache` variables carry the right logical-axis names for each `model_mode` (`CACHE_BATCH_PREFILL` vs `CACHE_BATCH`), which `bulk_insert` / `_insert_jit need` to match by substring.

Also: I re-ran the maxengine test with the latest code vs main branch and updated the link in the PR body.

**Review by bvandermoon — APPROVED** [review]

File: src/maxtext/inference/maxengine/maxengine.py [hunk 4]
```diff
@@ -132,6 +159,63 @@ def print_stats(self, label: str):
...
     max_utils.print_cpu_ram_stats(label)
 
+  # NNX cache adapter: bulk_insert / _insert_jit / _maybe_stack_* switch on
+  # path[-1].key (e.g. "cached_prefill_key"). NNX state would expose ".value" at
+  # that position, so we convert NNX state <-> plain dict at the JIT boundary
+  # via to_pure_dict / replace_by_pure_dict. The cache helpers stay unchanged.
+
+  def _nnx_cache_state_template(self, mode: str = MODEL_MODE_PREFILL) -> Any:
+    """Empty nnx.State template for the model's nnx.Cache vars (PREFILL=batch 1, AR=batch N)."""
+    src = self.model if mode == MODEL_MODE_PREFILL else self.model_ar
```
Comment (NuojCheng) [inline src/maxtext/inference/maxengine/maxengine.py]: we might need check/value error somewhere to make sure mode can only be prefill or ar

File: src/maxtext/inference/maxengine/maxengine.py [hunk 5]
```diff
@@ -110,10 +114,33 @@ def __init__(self, config: Any, devices: Any | None = None):
...
+      # We need both PREFILL and AR abstract models because the cache vars inherit
+      # CACHE_BATCH_PREFILL vs CACHE_BATCH from the construction model_mode, and
+      # bulk_insert searches for the substring "cache_batch" in the AR-mode names.
+      # Calling nnx.eval_shape directly (instead of create_nnx_abstract_model) avoids
+      # the jax.set_mesh wrap that trips Flax 0.12.6 on logical-only axes like "norm".
+      _create_model = model_creation_utils.get_nnx_create_model_fn(config, mesh=self._mesh, model_mode=MODEL_MODE_PREFILL)
+      _create_model_ar = model_creation_utils.get_nnx_create_model_fn(
+          config, mesh=self._mesh, model_mode=MODEL_MODE_AUTOREGRESSIVE
+      )
+      with nn_partitioning.axis_rules(config.logical_axis_rules):
```
Comment (NuojCheng) [inline src/maxtext/inference/maxengine/maxengine.py]: with jax.set_mesh(self.mesh), nn_partitioning...:

File: src/maxtext/inference/maxengine/maxengine.py [hunk 6]
```diff
@@ -1609,6 +1818,51 @@ def is_lp(k):
     zeroed = max_utils.unbox_logicallypartioned(init_state)
     return zeroed
 
+  def _init_decode_state_nnx(self, rng) -> DecodeState:
+    """NNX equivalent of init_decode_state. Returns a decode_state dict with a pure-dict cache."""
+    del rng  # cache shape comes from the abstract model
```
Comment (NuojCheng) [inline src/maxtext/inference/maxengine/maxengine.py]: I think XLA gonna ignore del in python right? It is really helpful?

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/inference/maxengine/maxengine.py (hunk 6 above)
Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py]: I removed the `rng` parameter from `_init_decode_state_nnx` entirely since the cache is zeros and doesn't need any randomness, and updated the caller.

File: src/maxtext/inference/maxengine/maxengine.py (hunk 5 above)
Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py]: Applied, thank you.

File: src/maxtext/inference/maxengine/maxengine.py (hunk 4 above)
Comment (ecnal-cienet) [reply src/maxtext/inference/maxengine/maxengine.py]: Hi @NuojCheng, I added a `_abstract_model_for_mode(mode)` helper that validates mode and raises `ValueError` if it's not `MODEL_MODE_PREFILL` or `MODEL_MODE_AUTOREGRESSIVE`. Both `_nnx_cache_state_template` and `_nnx_init_cache_dict` now go through it.
